client/redux/features/api/slice.ts

```diff
@@ -0,0 +1,54 @@
+import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
+import { RootState } from "@/redux/index";
+import type { LoginRequest, LoginResponse, User } from "@/redux/index";
+
+/*To interact with API endpoints, we need to create an API slice.
+ Application should have one createApi call, to the same base URL, that should contain all endpoint definitions
+*/
+
+// 🚀 Create an API instance
+export const apiSlice = createApi({
+  // 🌐 Configure the base query
+  baseQuery: fetchBaseQuery({
+    baseUrl: "http://localhost:3000/api/",
+    // 🛡️ Prepare headers for authenticated requests
+    prepareHeaders: (headers, { getState }) => {
+      // By default, if we have a token in the store, let's use that for authenticated requests
+      const token = (getState() as RootState).auth.token;
+      if (token) {
+        headers.set("authorization", `Bearer ${token}`);
+      }
+      return headers;
+    },
+  }),
+  // 🏁 Define API endpoints
+  /*
+  Each endpoint object contains:
+    The same primary query/mutation hook that we exported from the root API slice object, but named as useQuery or useMutation
+    For query endpoints, an additional set of query hooks for scenarios like "lazy queries" or partial subscriptions
+    A set of "matcher" utilities to check for the pending/fulfilled/rejected actions dispatched by requests for this endpoint
+    An initiate thunk that triggers a request for this endpoint
+    A select function that creates memoized selectors that can retrieve the cached result data + status entries for this endpoint
+  */
+  endpoints: (builder) => ({
+    // Define a login mutation
+    //                       ResultType    QueryArg
+    //                           v           v
+    login: builder.mutation<LoginResponse, LoginRequest>({
+      query: (credentials) => ({
+        url: "login",
+        method: "POST",
+        body: credentials,
+      }),
+    }),
+    // Define a users query
```

Comment:
We should do all our CRUD gets etc. from this api slice?

Reply:
Typically, one API slice **per base URL** to which the application communicates.
In this demo, I made a generically name apiSlice inferring only 1 base URL but, I suspect real use cases will require multiple slices for example authAPI, for crudAPI , for reportAPI etc.  each having a different base URL and appropriate namespace.